File: circularlayoutmanager/src/main/java/com/kapil/circularlayoutmanager/ScrollWheel.java
package com.kapil.circularlayoutmanager;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.util.AttributeSet;
import android.view.GestureDetector;
import android.view.MotionEvent;
import android.view.View;

/**
 * This view implements scroll wheel functionality.
 *
 * It converts circular touch motion into feed for the scrolling of a recycler view.
 */

public class ScrollWheel extends View implements GestureDetector.OnGestureListener {
    private GestureDetector gestureDetector;
    private RecyclerView recyclerView;

    private OnItemClickListener onItemClickListener;

    private boolean scrollWheelEnabled;
    private boolean consumeTouchOutsideTouchAreaEnabled;

    private int xCenter;
    private int yCenter;
    private int innerRadius;
    private int outerRadius;
    private int touchAreaThickness;
    private boolean touchInitiatedBetweenCircles;

    private Paint touchAreaPaint;
    private boolean highlightTouchAreaEnabled;

    public ScrollWheel(Context context) {
        super(context);
        init();
    }

    public ScrollWheel(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public ScrollWheel(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    private void init() {
        gestureDetector = new GestureDetector(getContext(), this);

        scrollWheelEnabled = true;
        consumeTouchOutsideTouchAreaEnabled = true;

        touchAreaThickness = -1;

        touchAreaPaint = new Paint();
        touchAreaPaint.setAntiAlias(true);
        touchAreaPaint.setStyle(Paint.Style.STROKE);
        touchAreaPaint.setColor(Color.parseColor("#50FF0000"));

        highlightTouchAreaEnabled = true;
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);

        xCenter = getMeasuredWidth() / 2;
        yCenter = getMeasuredHeight() / 2;

        int minDimension = Math.min(getMeasuredWidth(), getMeasuredHeight());


        outerRadius = minDimension / 2;
        innerRadius = touchAreaThickness == -1 ? minDimension / 4 : outerRadius - touchAreaThickness;

        touchAreaPaint.setStrokeWidth(outerRadius - innerRadius);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        if (highlightTouchAreaEnabled) {
            canvas.drawCircle(xCenter, yCenter, (innerRadius + outerRadius) / 2, touchAreaPaint);
        }
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        if ((recyclerView == null) || (!scrollWheelEnabled)) {
            return super.onTouchEvent(event);
        }

        if (event.getActionMasked() == MotionEvent.ACTION_UP) {
            ((CircularLayoutManager) recyclerView.getLayoutManager()).stabilize();
        }

        return gestureDetector.onTouchEvent(event);
    }

    @Override
    public boolean onDown(MotionEvent e) {

        float x = e.getX();
        float y = e.getY();

        touchInitiatedBetweenCircles = ((((x - xCenter) * (x - xCenter)) + ((y - yCenter) * (y - yCenter))) > (innerRadius * innerRadius))
                && ((((x - xCenter) * (x - xCenter)) + ((y - yCenter) * (y - yCenter))) < (outerRadius * outerRadius));

        return consumeTouchOutsideTouchAreaEnabled || touchInitiatedBetweenCircles;
    }

    @Override
    public void onShowPress(MotionEvent e) {

    }

    @Override
    public boolean onSingleTapUp(MotionEvent e) {
        if (onItemClickListener != null) {
            int childIndex = getChildIndexUnder(e.getX(), e.getY());
            if (childIndex != -1) {
                onItemClickListener.onItemClick(ScrollWheel.this, childIndex);
            }
        }
        return true;
    }

    @Override
    public boolean onScroll(MotionEvent e1, MotionEvent e2, float distanceX, float distanceY) {
        if (!touchInitiatedBetweenCircles) {
            return false;
        }

        int delta = 0;

        float x = e2.getX();
        float y = e2.getY();

        if ((x <= xCenter) && (y < yCenter)) {
            delta = (int) (distanceX - distanceY);
        } else if ((x > xCenter) && (y <= yCenter)) {
            delta = (int) (distanceX + distanceY);
        } else if ((x >= xCenter) && (y > yCenter)) {
            delta = (int) (-distanceX + distanceY);
        } else if ((x < xCenter) && (y >= yCenter)) {
            delta = (int) (-distanceX - distanceY);
        }

        recyclerView.scrollBy(0, delta);

        return true;
    }

    @Override
    public void onLongPress(MotionEvent e) {
        if ((onItemClickListener != null) && (consumeTouchOutsideTouchAreaEnabled)) {
            int childIndex = getChildIndexUnder(e.getX(), e.getY());
            if (childIndex != -1) {
                onItemClickListener.onItemLongClick(ScrollWheel.this, childIndex);
            }
        }
    }

    @Override
    public boolean onFling(MotionEvent e1, MotionEvent e2, float velocityX, float velocityY) {
        if (!touchInitiatedBetweenCircles) {
            return false;
        }

        int delta = 0;

        float x = e2.getX();
        float y = e2.getY();

        if ((x <= xCenter) && (y < yCenter)) {
            delta = (int) (velocityX - velocityY);
        } else if ((x > xCenter) && (y <= yCenter)) {
            delta = (int) (velocityX + velocityY);
        } else if ((x >= xCenter) && (y > yCenter)) {
            delta = (int) (-velocityX + velocityY);
        } else if ((x < xCenter) && (y >= yCenter)) {
            delta = (int) (-velocityX - velocityY);
        }

        recyclerView.fling(0, -delta);

        return true;
    }

    int getChildIndexUnder(float x, float y) {
        View child = recyclerView.findChildViewUnder(x, y);
        if (child != null) {
            return recyclerView.indexOfChild(child);
        }
        return -1;
    }

    public RecyclerView getRecyclerView() {
        return recyclerView;
    }

    public void setRecyclerView(RecyclerView recyclerView) {
        this.recyclerView = recyclerView;
    }

    public void setOnItemClickListener(OnItemClickListener onItemClickListener) {
        this.onItemClickListener = onItemClickListener;
    }

    public boolean isScrollWheelEnabled() {
        return scrollWheelEnabled;
    }

    public void setScrollWheelEnabled(boolean scrollWheelEnabled) {
        this.scrollWheelEnabled = scrollWheelEnabled;
    }

    public boolean isConsumeTouchOutsideTouchAreaEnabled() {
        return consumeTouchOutsideTouchAreaEnabled;
    }

    public void setConsumeTouchOutsideTouchAreaEnabled(boolean consumeTouchOutsideTouchAreaEnabled) {
        this.consumeTouchOutsideTouchAreaEnabled = consumeTouchOutsideTouchAreaEnabled;
    }

    public int getTouchAreaThickness() {
        return touchAreaThickness;
    }

    public void setTouchAreaThickness(int thickness) {
        touchAreaThickness = (int) Utils.dpToPx(getContext(), thickness);
        invalidate();
    }

    public int getTouchAreaColor() {
        return touchAreaPaint.getColor();
    }

    public void setTouchAreaColor(int touchAreaColor) {
        touchAreaPaint.setColor(touchAreaColor);
        invalidate();
    }

    public boolean isHighlightTouchAreaEnabled() {
        return highlightTouchAreaEnabled;
    }

    public void setHighlightTouchAreaEnabled(boolean highlightTouchAreaEnabled) {
        this.highlightTouchAreaEnabled = highlightTouchAreaEnabled;
        invalidate();
    }
}

File: circular-layout-manager/src/main/java/com/kapil/circularlayoutmanager/ScrollWheel.kt
package com.kapil.circularlayoutmanager

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.util.AttributeSet
import android.view.GestureDetector
import android.view.MotionEvent
import android.view.View
import androidx.annotation.ColorInt
import androidx.annotation.Px
import androidx.core.content.ContextCompat
import kotlin.math.min

/**
 * This view implements scroll wheel functionality.
 *
 * It converts circular touch motion into input for the scrolling of a recycler view.
 *
 * Touch Area of the scroll wheel is defined as the region between the circles formed by the outer
 * and inner radii. Circular scrolling could be initiated only from this region.
 */
class ScrollWheel @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : View(context, attrs, defStyleAttr), GestureDetector.OnGestureListener {

    /**
     * Function to show or hide path of action of the scroll wheel. Enabling it will highlight an
     * area on the screen as a cue for the user to use the scroll wheel.
     *
     * By default set to true.
     */
    var isHighlightTouchAreaEnabled = true
        set(value) {
            field = value
            invalidate()
        }

    /**
     * Toggle to enable or disable detection of clicks and long clicks.
     *
     * It could be set to true if the touch area lies over the list partially or fully, and to false
     * otherwise.
     *
     * By default set to true.
     */
    var isHandleClicksEnabled: Boolean = true

    @Px
    var touchAreaThickness =
        context.resources.getDimension(R.dimen.default_touch_area_thickness)
        set(value) {
            field = value
            invalidate()
        }

    var onItemClickListener: ((Float, Float) -> Unit)? = null
    var onItemLongClickListener: ((Float, Float) -> Unit)? = null

    /**
     * The scroll and fling listeners return the distance and velocity respectively by which the
     * user has scrolled or flung. A positive value indicates movement in anticlockwise direction
     * and a negative value indicates a movement in clockwise direction.
     */
    var onScrollListener: ((Float) -> Unit)? = null
    var onFlingListener: ((Float) -> Unit)? = null

    var onTouchReleasedListener: (() -> Unit)? = null

    private var xCenter = 0f
    private var yCenter = 0f
    private var innerRadius = 0f
    private var outerRadius = 0f

    private var touchInitiatedBetweenCircles = false

    private var touchAreaPaint: Paint = Paint().apply {
        isAntiAlias = true
        style = Paint.Style.STROKE
        color = ContextCompat.getColor(context, R.color.default_touch_area_color)
    }

    private var gestureDetector: GestureDetector = GestureDetector(context, this)

    init {
        context.theme.obtainStyledAttributes(attrs, R.styleable.ScrollWheel, 0, 0).apply {
            isHandleClicksEnabled = getBoolean(R.styleable.ScrollWheel_isHandleClicksEnabled, true)
            isHighlightTouchAreaEnabled =
                getBoolean(R.styleable.ScrollWheel_isHighlightTouchAreaEnabled, true)
            touchAreaThickness = getDimension(
                R.styleable.ScrollWheel_touchAreaThickness,
                context.resources.getDimension(R.dimen.default_touch_area_thickness)
            )
            setTouchAreaColor(
                getColor(
                    R.styleable.ScrollWheel_touchAreaColor,
                    ContextCompat.getColor(context, R.color.default_touch_area_color)
                )
            )
            recycle()
        }
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        xCenter = measuredWidth / 2f
        yCenter = measuredHeight / 2f
        outerRadius = min(measuredWidth, measuredHeight) / 2f
        innerRadius = outerRadius - touchAreaThickness
        touchAreaPaint.strokeWidth = outerRadius - innerRadius
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        if (isHighlightTouchAreaEnabled)
            canvas.drawCircle(xCenter, yCenter, (innerRadius + outerRadius) / 2f, touchAreaPaint)
    }

    // As the functionality of the view revolves around just providing a different experience for
    // scrolling and not enabling scrolling itself, implementing accessibility features would
    // provide no added value
    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(event: MotionEvent): Boolean {
        if (!isEnabled) return super.onTouchEvent(event)
        if (event.actionMasked == MotionEvent.ACTION_UP) onTouchReleasedListener?.invoke()
        return gestureDetector.onTouchEvent(event)
    }

    override fun onDown(e: MotionEvent): Boolean {
        val x = e.x
        val y = e.y
        touchInitiatedBetweenCircles =
            ((x - xCenter) * (x - xCenter) + (y - yCenter) * (y - yCenter) > innerRadius * innerRadius
                    && (x - xCenter) * (x - xCenter) + (y - yCenter) * (y - yCenter) < outerRadius * outerRadius)
        return isHandleClicksEnabled || touchInitiatedBetweenCircles
    }

    override fun onScroll(
        e1: MotionEvent,
        e2: MotionEvent,
        distanceX: Float,
        distanceY: Float
    ): Boolean {
        if (touchInitiatedBetweenCircles)
            onScrollListener?.invoke(calculateDelta(e2.x, e2.y, distanceX, distanceY))
        return touchInitiatedBetweenCircles
    }

    override fun onFling(
        e1: MotionEvent,
        e2: MotionEvent,
        velocityX: Float,
        velocityY: Float
    ): Boolean {
        if (touchInitiatedBetweenCircles)
            onFlingListener?.invoke(-calculateDelta(e2.x, e2.y, velocityX, velocityY))
        return touchInitiatedBetweenCircles
    }

    override fun onSingleTapUp(e: MotionEvent): Boolean {
        if (isHandleClicksEnabled) onItemClickListener?.invoke(e.x, e.y)
        return isHandleClicksEnabled
    }

    override fun onLongPress(e: MotionEvent) {
        if (isHandleClicksEnabled) onItemLongClickListener?.invoke(e.x, e.y)
    }

    override fun onShowPress(e: MotionEvent) {}

    private fun calculateDelta(x: Float, y: Float, dx: Float, dy: Float) = when {
        (x <= xCenter && y < yCenter) -> (dx - dy)
        (x > xCenter && y <= yCenter) -> (dx + dy)
        (x >= xCenter && y > yCenter) -> (-dx + dy)
        (x < xCenter && y >= yCenter) -> (-dx - dy)
        else -> 0f
    }

    @ColorInt
    fun getTouchAreaColor() = touchAreaPaint.color

    fun setTouchAreaColor(@ColorInt touchAreaColor: Int) {
        touchAreaPaint.color = touchAreaColor
        invalidate()
    }
}